promql/engine.go

```diff
@@ -2141,3 +2253,163 @@ func unwrapParenExpr(e *parser.Expr) {
+// passed expression qualifies to be wrapped by StepInvariantExpr.
+func wrapWithStepInvariantExprHelper(expr parser.Expr) bool {
+	switch n := expr.(type) {
+	case *parser.VectorSelector:
+		return n.Timestamp != nil
+
+	case *parser.AggregateExpr:
+		return wrapWithStepInvariantExprHelper(n.Expr)
+
+	case *parser.BinaryExpr:
+		isInvariant1, isInvariant2 := wrapWithStepInvariantExprHelper(n.LHS), wrapWithStepInvariantExprHelper(n.RHS)
+		if isInvariant1 && isInvariant2 {
+			return true
+		}
+
+		if isInvariant1 {
+			n.LHS = newStepInvariantExpr(n.LHS)
+		}
+		if isInvariant2 {
+			n.RHS = newStepInvariantExpr(n.RHS)
+		}
+
+		return false
+
+	case *parser.Call:
+		isStepInvariant := IsFunctionStepInvariant(n)
+		isStepInvariantSlice := make([]bool, len(n.Args))
+		for i := range n.Args {
+			isStepInvariantSlice[i] = wrapWithStepInvariantExprHelper(n.Args[i])
+			isStepInvariant = isStepInvariant && isStepInvariantSlice[i]
+		}
+
+		if n.Func.Name == "timestamp" {
+			// If timestamp(..) has something other than vector selector inside,
+			// then it is not step invariant.
+			_, ok := n.Args[0].(*parser.VectorSelector)
+			isStepInvariant = isStepInvariant && ok
+		}
+		if isStepInvariant {
+
+			// The function and all arguments are step invariant.
+			return true
+		}
+
+		for i, ic := range isStepInvariantSlice {
+			if ic {
+				n.Args[i] = newStepInvariantExpr(n.Args[i])
+			}
+		}
+		return false
+
+	case *parser.MatrixSelector:
+		return n.VectorSelector.(*parser.VectorSelector).Timestamp != nil
+
+	case *parser.SubqueryExpr:
+		// Since we adjust offset for the @ modifier evaluation,
+		// it gets tricky to adjust it for every subquery step.
+		// Hence we wrap the inside of subquery irrespective of
+		// @ on subquery (given it is also step invariant) so that
+		// it is evaluated only once w.r.t. the start time of subquery.
+		isInvariant := wrapWithStepInvariantExprHelper(n.Expr)
+		if isInvariant {
+			n.Expr = newStepInvariantExpr(n.Expr)
+		}
+		return n.Timestamp != nil
+
+	case *parser.ParenExpr:
+		return wrapWithStepInvariantExprHelper(n.Expr)
+
+	case *parser.UnaryExpr:
+		return wrapWithStepInvariantExprHelper(n.Expr)
+
+	case *parser.StringLiteral, *parser.NumberLiteral:
+		return true
+	}
+
+	panic(fmt.Sprintf("found unexpected node %#v", expr))
+}
+
+func newStepInvariantExpr(expr parser.Expr) parser.Expr {
+	switch e := expr.(type) {
+	case *parser.ParenExpr:
+		// Wrapping the inside of () makes it easy to unwrap the paren later.
+		return newStepInvariantExpr(e.Expr)
+	}
+	if dontWrapStepInvariantExpr(expr) {
+		return expr
+	}
+	return &parser.StepInvariantExpr{Expr: expr}
+}
+
+// dontWrapStepInvariantExpr returns true if we should not wrap an expression even
+// if it step invariant. That includes constant literals like numbers and string where we don't
+// need any kind special handling for result hence reduce engine complexity in handling them.
```

Comment:
reduce the

I'm not seeing how this is the case, wouldn't they still be handled correctly if wrapped? All the code still exists one way or the other.

Reply:
literals are not handled properly unfortunately if they are wrapped, most likely the timestamp of the number in play?

Reply:
Scalars don't have timestamps, so something weird is going on here that needs investigation.

Reply:
Apparently the string literal was returning `type String struct { T int64, V string }` with a method `func (s String) MarshalJSON() ([]byte, error) { return json.Marshal([...]interface{}{float64(s.T) / 1000, s.V}) }`, I am not aware of the significance of `MarshalJSON`, I think I can remove it? Removing it and removing this dont wrap works.

Reply:
A string literal on its own is a valid PromQL expression, if you remove the marshalling I'm pretty sure you'll break things. The marshalling should be irrelevant here though, as it'd only affect the API.

Reply:
:facepalm: right, I don't remember now what was failing before. Added back the timestamp and marshalling, all works!